File: scripts/exposure-auc.py
"""
DOES CONDITIONING ON THE AUDIENCE BEAT A GLOBAL VOTE COUNT?

    python3 scripts/exposure-auc.py

The honest test of an exposure prior, and the one this project has never run.

For each held-out person: take half their viewing history, work out which taste
communities they belong to, and use that to predict the *other* half against
films they never watched. Score it as AUC — the probability that a film they
did watch is ranked above one they did not.

Three predictors compete on identical data:

    votes      TMDB vote count, what ships today
    global     how widely MovieLens watched it, ignoring who is asking
    community  posterior-weighted reach across taste communities

`global` is in here deliberately. Without it a win for `community` could just be
"MovieLens knows more about watching than TMDB does", which would be true and
would not justify a single line of the clustering.

HOLDOUT. The evaluated users are excluded from the clustering entirely — they
contribute to no centroid and to no community's reach figure. Five separate
times in this project an instrument has been fitted on the thing it was
measuring, so the disjointness here is enforced in code rather than assumed.
"""

import base64
import csv
import json
import os
import time
from collections import defaultdict

import numpy as np

ML = ".cache/ml-32m"
DIM = int(os.environ.get("DIM", 96))
K = int(os.environ.get("K", 100))
ITERS = int(os.environ.get("ITERS", 12))
MIN_RATINGS = int(os.environ.get("MIN_RATINGS", 40))
HELD_OUT = int(os.environ.get("HELD_OUT", 4000))
NEGATIVES = int(os.environ.get("NEGATIVES", 200))

t0 = time.time()


def log(m):
    print(f"  [{time.time() - t0:6.1f}s] {m}", flush=True)


cat = json.load(open("public/catalog.json"))
rows = cat["t"]
ids = [f"{'movie' if r[1] == 0 else 'tv'}-{r[0]}" for r in rows]
# the encoded catalog keeps vote count at a fixed offset; find it by shape
votes = {}
for r, tid in zip(rows, ids):
    nums = [x for x in r if isinstance(x, (int, float))]
    votes[tid] = r
cat_index = {t: i for i, t in enumerate(ids)}

tmdb_to_ours = {t[6:]: t for t in ids if t.startswith("movie-")}
ml_to_ours = {}
with open(f"{ML}/links.csv") as f:
    for row in csv.DictReader(f):
        t = (row.get("tmdbId") or "").strip()
        if t and t in tmdb_to_ours:
            ml_to_ours[int(row["movieId"])] = tmdb_to_ours[t]

title_index, col_to_id = {}, {}
for mid, oid in ml_to_ours.items():
    if oid not in title_index:
        col_to_id[len(title_index)] = oid
        title_index[oid] = len(title_index)
n_titles = len(title_index)
ml_to_col = {m: title_index[ml_to_ours[m]] for m in ml_to_ours}
log(f"{n_titles} of our films are in MovieLens")

user_items = {}
with open(f"{ML}/ratings.csv") as f:
    next(f)
    cur, cols = None, []
    for line in f:
        u, m, _ = line.split(",", 2)
        c = ml_to_col.get(int(m))
        if c is None:
            continue
        if u != cur:
            if cur is not None and len(cols) >= MIN_RATINGS:
                user_items[cur] = cols
            cur, cols = u, []
        cols.append(c)
    if cur is not None and len(cols) >= MIN_RATINGS:
        user_items[cur] = cols
users = list(user_items)
log(f"{len(users)} users with {MIN_RATINGS}+ of our films")

rng = np.random.default_rng(7)
order = rng.permutation(len(users))
test_users = [users[i] for i in order[:HELD_OUT]]
train_users = [users[i] for i in order[HELD_OUT:]]
log(f"held out {len(test_users)} users; {len(train_users)} build the communities")

# ── communities, from training users only ────────────────────────────────
sig = rng.choice(np.array([-1.0, 1.0], dtype=np.float32), size=(n_titles, DIM))
V = np.zeros((len(train_users), DIM), dtype=np.float32)
for i, u in enumerate(train_users):
    V[i] = sig[user_items[u]].mean(axis=0)
V /= np.maximum(np.linalg.norm(V, axis=1, keepdims=True), 1e-9)

C = V[rng.choice(len(train_users), size=K, replace=False)].copy()
for _ in range(ITERS):
    assign = np.empty(len(train_users), dtype=np.int32)
    for s in range(0, len(train_users), 200_000):
        assign[s : s + 200_000] = np.argmax(V[s : s + 200_000] @ C.T, axis=1)
    for k in range(K):
        mem = V[assign == k]
        if len(mem) == 0:
            C[k] = V[rng.integers(len(train_users))]
            continue
        c = mem.mean(axis=0)
        C[k] = c / max(np.linalg.norm(c), 1e-9)

counts = np.zeros((K, n_titles), dtype=np.int32)
for i, u in enumerate(train_users):
    counts[assign[i], user_items[u]] += 1
members = np.maximum(np.bincount(assign, minlength=K), 1).astype(np.float32)
reach = counts / members[:, None]
glob = counts.sum(axis=0) / float(len(train_users))
log("communities built")

# ── the three predictors ─────────────────────────────────────────────────
vote_of = np.zeros(n_titles, dtype=np.float32)
for c in range(n_titles):
    tid = col_to_id[c]
    r = rows[cat_index[tid]]
    # vote count is the last large integer on the row; take the max integer
    ints = [x for x in r if isinstance(x, int)]
    vote_of[c] = max(ints[2:]) if len(ints) > 2 else 0

LOG_R = np.log(np.clip(reach, 1e-4, 1 - 1e-4))
LOG_NR = np.log(np.clip(1 - reach, 1e-4, 1 - 1e-4))

aucs = {"votes": [], "global": [], "community": []}
for u in test_users:
    items = user_items[u]
    if len(items) < MIN_RATINGS:
        continue
    half = len(items) // 2
    fit, hidden = items[:half], items[half:]
    seen_all = set(items)
    negs = []
    while len(negs) < NEGATIVES:
        c = int(rng.integers(n_titles))
        if c not in seen_all:
            negs.append(c)

    # posterior over communities from the fitting half only
    lp = LOG_R[:, fit].sum(axis=1)
    lp -= lp.max()
    post = np.exp(lp)
    post /= post.sum()
    comm_score = post @ reach

    for name, score in (
        ("votes", vote_of),
        ("global", glob),
        ("community", comm_score),
    ):
        pos = score[hidden]
        neg = score[negs]
        wins = (pos[:, None] > neg[None, :]).sum() + 0.5 * (pos[:, None] == neg[None, :]).sum()
        aucs[name].append(wins / (len(pos) * len(neg)))

log(f"scored {len(aucs['votes'])} held-out people")
print("\n  AUC — probability a film they watched outranks one they did not\n")
for name in ("votes", "global", "community"):
    a = np.array(aucs[name])
    print(f"    {name:12} {a.mean():.4f}   +/- {1.96 * a.std() / np.sqrt(len(a)):.4f}")
print("\n  0.5 is a coin flip. `votes` is what ships today.\n")

File: scripts/exposure-auc.ts
/**
 * DOES THE ENGINE KNOW WHAT HE HAS WATCHED? MEASURED, ON DATA IT DID NOT PICK.
 *
 *   npx tsx scripts/exposure-auc.ts
 *
 * `watchLikelihood` is the one model that decides whether the product's goal is
 * reachable. Every card the deck deals is a bet on it, and the goal — get a
 * person's whole watch history into the site inside a week — is nothing but
 * that bet repeated a few thousand times.
 *
 * It has never been measured on honest data. The one number this project has,
 * AUC 0.500 for the fame prior, came off a sample truncated by fame, where
 * range restriction drives AUC to 0.5 mechanically. That measurement could not
 * have come out any other way, so it says nothing.
 *
 * This one can. Two sources, and the split between them is the whole design:
 *
 *   TRAIN   1,826 titles the deck dealt him across six sessions, with his
 *           answers. Biased by construction — the engine chose every one.
 *   TEST    599 titles drawn at random from every depth of the catalog by
 *           `/calibrate`, with his answers. Nothing here was ranked, gated,
 *           scored or personalised. **The engine did not choose this set.**
 *
 * Training on the biased set and testing on the unbiased one is the only way
 * to ask "does this model generalise beyond the region it was fitted in",
 * which is precisely the question the decaying tail of his session poses.
 *
 * Titles present in both are dropped from **training**, not from the test set:
 * scoring the model on answers it was trained on measures memory rather than
 * prediction, and dropping them from the test instead would throw away
 * two-thirds of the positives — 30 of them is already a thin base.
 *
 * AUC here is the probability that a title he HAS watched scores above one he
 * has NOT, drawn at random. 0.5 is a coin flip. It is the right measure because
 * it is invariant to the base rate, and his base rate is 5%.
 */
import { readFileSync, existsSync } from "node:fs";
import { decodeCatalog, type EncodedCatalog } from "../src/lib/data/catalog-codec";
import { buildRarityIndex, titleTokens } from "../src/lib/engine/facets";
import { featurize } from "../src/lib/engine/features";
import { applySwipe, emptyProfile, watchLikelihood, seenTrust } from "../src/lib/engine/taste";
import type { SwipeAction, Title } from "../src/lib/types";
import { loadFullCatalog } from "./lib/catalog";

const catalog = loadFullCatalog();
buildRarityIndex(catalog);
const byId = new Map(catalog.map((t) => [t.id, t]));

/** the fame prior exactly as `recommend` computes it, so this grades what ships */
const ranked = [...catalog].sort((a, b) => b.voteCount - a.voteCount);
const fameRank = new Map(ranked.map((t, i) => [t.id, i + 1]));

type Row = { id: string; a: SwipeAction; at: number };
type Cal = { id: string; seen: boolean };

const SESSIONS = [
  ".cache/user-swipes.json",
  ".cache/user-swipes-v2.json",
  ".cache/user-swipes-v3.json",
  ".cache/user-swipes-v4.json",
  ".cache/user-swipes-v5.json",
  ".cache/user-swipes-v6.json",
].filter(existsSync);

const CAL = [
  ".cache/calibration-199.json",
  ".cache/calibration-200-v2.json",
  ".cache/calibration-200-v3.json",
].filter(existsSync);

/* ── test: the blind sample, whole ── */
const cal: Cal[] = CAL.flatMap((f) => (JSON.parse(readFileSync(f, "utf8")) as { sample: Cal[] }).sample);
const test = cal.filter((r) => byId.has(r.id));
const testIds = new Set(test.map((r) => r.id));

/* ── train: every answer he has ever given a card, minus the test set ── */
const latest = new Map<string, { action: SwipeAction; at: number }>();
for (const f of SESSIONS) {
  const raw = JSON.parse(readFileSync(f, "utf8")) as { swipes?: Row[] } | Row[];
  for (const r of Array.isArray(raw) ? raw : (raw.swipes ?? [])) {
    const prev = latest.get(r.id);
    if (!prev || r.at > prev.at) latest.set(r.id, { action: r.a, at: r.at });
  }
}
const train: { title: Title; action: SwipeAction; at: number }[] = [];
let overlap = 0;
for (const [id, v] of latest) {
  if (testIds.has(id)) {
    overlap++;
    continue;
  }
  const t = byId.get(id);
  if (t) train.push({ title: t, action: v.action, at: v.at });
}
train.sort((a, b) => a.at - b.at);

const pos = test.filter((r) => r.seen).length;
console.log(
  `\nTRAIN  ${train.length} titles the deck dealt him across six sessions,` +
    `\n       after removing ${overlap} that also appear in the blind sample`
);
console.log(`TEST   ${test.length} titles from the blind sample — the engine chose none of them`);
console.log(`       of which ${pos} he has watched (${((100 * pos) / test.length).toFixed(1)}%)\n`);

/** area under the ROC curve, by rank-sum — no binning, no threshold */
function auc(scored: { score: number; label: boolean }[]): number {
  const sorted = [...scored].sort((a, b) => a.score - b.score);
  let rank = 0;
  let sumPos = 0;
  let nPos = 0;
  for (let i = 0; i < sorted.length; ) {
    let j = i;
    while (j < sorted.length && sorted[j].score === sorted[i].score) j++;
    const avgRank = (i + j + 1) / 2; // 1-based, ties share the average
    for (let k = i; k < j; k++) {
      rank++;
      if (sorted[k].label) {
        sumPos += avgRank;
        nPos++;
      }
    }
    i = j;
  }
  const nNeg = sorted.length - nPos;
  if (nPos === 0 || nNeg === 0) return 0.5;
  return (sumPos - (nPos * (nPos + 1)) / 2) / (nPos * nNeg);
}

/**
 * Build the profile from the first `n` training answers, then score the whole
 * held-out set. Sweeping n shows whether more of his answers make the model
 * better — which is the claim the whole personalised gate rests on.
 */
function evaluate(n: number) {
  let profile = emptyProfile();
  for (const r of train.slice(0, n)) {
    profile = applySwipe(profile, r.title, featurize(r.title), r.action);
  }
  const trust = seenTrust(profile);
  const rows = test.map((r) => {
    const t = byId.get(r.id)!;
    const rank = fameRank.get(r.id)!;
    /* the same shape `recommend` uses: a fame prior in [0,1], famous = high */
    const fame = 1 - rank / catalog.length;
    return {
      fame,
      model: watchLikelihood(profile, titleTokens(t), fame),
      label: r.seen,
    };
  });
  return {
    n,
    trust,
    fame: auc(rows.map((r) => ({ score: r.fame, label: r.label }))),
    model: auc(rows.map((r) => ({ score: r.model, label: r.label }))),
  };
}

console.log("  answers trained on   seenTrust    AUC fame only    AUC watchLikelihood");
for (const n of [0, 50, 100, 200, 400, 800, 1200, train.length]) {
  if (n > train.length) continue;
  const r = evaluate(n);
  const delta = r.model - r.fame;
  console.log(
    `  ${String(r.n).padStart(18)}   ${r.trust.toFixed(3).padStart(8)}    ` +
      `${r.fame.toFixed(3).padStart(12)}    ${r.model.toFixed(3).padStart(18)}` +
      `   ${delta >= 0 ? "+" : ""}${delta.toFixed(3)}`
  );
}

console.log(
  `\n  AUC 0.5 is a coin flip. Above 0.5 means a title he has watched tends to` +
    `\n  score above one he has not — which is the only thing the deck needs.\n`
);

/* ══════════════════════════════════════════════════════════════════════════
 * Every exposure signal in the catalog, on the same unbiased set.
 *
 * The blend above is worth +0.011 over fame. Before trying to improve the
 * blend it is worth knowing whether any *other* column predicts better than
 * the vote count does, because if one does then the engine is built on the
 * wrong prior and no amount of blending fixes that.
 *
 * `reach` is here because it was commissioned specifically to replace the vote
 * count, on the strength of the AUC 0.500 finding that this file has just
 * shown to be an artefact. It costs nothing to ask whether it was needed.
 * ═════════════════════════════════════════════════════════════════════════ */
let profile = emptyProfile();
for (const r of train) profile = applySwipe(profile, r.title, featurize(r.title), r.action);

/** how strongly a title is linked, by TMDB co-watch edges, to what he has watched */
const watchedIds = new Set(train.filter((r) => r.action !== "not_seen").map((r) => r.title.id));
const coWatch = (t: Title) => (t.related ?? []).filter((id) => watchedIds.has(id)).length;
/** and the reverse direction, which is the one that carries most of the graph */
const inbound = new Map<string, number>();
for (const t of catalog) {
  if (!watchedIds.has(t.id)) continue;
  for (const id of t.related ?? []) inbound.set(id, (inbound.get(id) ?? 0) + 1);
}

const signals: [string, (t: Title) => number][] = [
  ["vote count (what ships)", (t) => t.voteCount],
  ["reach (LLM estimate)", (t) => t.reach ?? -1],
  ["popularity", (t) => t.popularity],
  ["rating", (t) => t.rating],
  ["personal seen-facets only", (t) => seenScoreOf(t)],
  ["co-watch edges out", (t) => coWatch(t)],
  ["co-watch edges in", (t) => inbound.get(t.id) ?? 0],
  ["co-watch, either way", (t) => coWatch(t) + (inbound.get(t.id) ?? 0)],
];

function seenScoreOf(t: Title): number {
  /* watchLikelihood with the prior held flat isolates the personal half */
  return watchLikelihood(profile, titleTokens(t), 0);
}

/**
 * THIRTY POSITIVES. SAY THE INTERVAL OR SAY NOTHING.
 *
 * This project has tuned against a ruler that could not resolve the difference
 * it was being asked about five separate times. The test set here has 30
 * titles he has watched, and an AUC gap of 0.03 on 30 positives is well inside
 * the noise. So every number below carries a bootstrap interval, resampling
 * the test set 2,000 times, and any comparison whose intervals overlap is not
 * a finding — it is a coin landing.
 */
function bootstrapAuc(rows: { score: number; label: boolean }[], draws = 2000) {
  const out: number[] = [];
  let s = 12345;
  const rnd = () => ((s = (s * 1664525 + 1013904223) >>> 0), s / 0x100000000);
  for (let d = 0; d < draws; d++) {
    const pick: typeof rows = [];
    for (let i = 0; i < rows.length; i++) pick.push(rows[Math.floor(rnd() * rows.length)]);
    if (pick.some((r) => r.label) && pick.some((r) => !r.label)) out.push(auc(pick));
  }
  out.sort((a, b) => a - b);
  return [out[Math.floor(out.length * 0.025)], out[Math.floor(out.length * 0.975)]] as const;
}

console.log("  EVERY EXPOSURE SIGNAL, SAME UNBIASED TEST SET");
console.log("    signal                        AUC    95% interval     covers");
for (const [name, fn] of signals) {
  const rows = test.map((r) => ({ score: fn(byId.get(r.id)!), label: r.seen }));
  const usable = rows.filter((r) => r.score >= 0);
  const covered = usable.length;
  if (covered === 0) {
    console.log(`    ${name.padEnd(28)}   —    absent from the catalog        0%`);
    continue;
  }
  const a = auc(usable);
  const [lo, hi] = bootstrapAuc(usable);
  console.log(
    `    ${name.padEnd(28)} ${a.toFixed(3)}   ${lo.toFixed(3)} – ${hi.toFixed(3)}     ` +
      `${((100 * covered) / rows.length).toFixed(0)}%` +
      (name.startsWith("vote") ? "   <- the prior the gate is built on" : "")
  );
}
console.log(
  `\n  Intervals that overlap are not a ranking. On ${pos} positives, only a gap` +
    `\n  wider than about 0.10 is worth acting on.\n`
);
